rfc/openfed__prerequisite.md

```diff
@@ -0,0 +1,408 @@
+- Validate an input against another operation.
+
+Potential idea and discussion point:
+- Inject input through the response of an exterior query.
+
+# Definition
+This is the proposed definition of the directive (functionality is discussed in the appropriate section):
+```graphql
+directive @openfed__prerequisite(
+  resolveEntity: openfed__ResolveEntityInput
+  resolveQuery: openfed__ResolveQueryInput
+) on ARGUMENT_DEFINITION
+
+input openfed__ResolveEntityInput {
+  typeName: String!
+  fields: openfed__InputSet!
+}
+
+input openfed__ResolveQueryInput {
+  query: String!
+}
+
+scalar openfed__InputSet
+```
+
+# Functionality
+
+The `prerequisite` directive (currently) defines two optional arguments.
+The arguments are mutually exclusive; and although they are both optional (limitation of GraphQL), one (and only one) 
+argument _must_ be defined.
+Failure to define an argument, or defining both, will result in a composition error.
+
+## 1. resolveEntity argument
+The `resolveEntity` argument requires two inputs:
+- `typeName`: The `__typename` of the entity to be validated.
+- `fields`: The key field(s) that form the primary key.
+
+### The typeName input
+The `typeName` input corresponds to the `__typename` of an entity that defines an entity resolver in the same subgraph
+that defines the directive.
+
+Composition will fail if the type referenced by the `typeName` input is not an Object that defines a valid `@key`
+directive with the `resolvable` argument either omitted or explicitly set to true.
+
+### The fields input
+The `fields` input is type `openfed__InputSet!` (henceforth simply `InputSet`).
+The `InputSet` Scalar is similar but distinct from the `openfed__FieldSet` Scalar, which is used to define the `fields`
+argument for a `@key` directive.
+
+This Scalar encodes two things:
+- The key fields for the entity.
+- The name of the Inputs whose value will be propagated to the corresponding key field.
+
+#### Directive defined on an argument that returns a leaf (Scalar or Enum)
+In the event that the directive is defined directly on an argument that returns a leaf type (Scalar or Enum):
+- `resolveEntity.fields` must define a single valid key field name. If the entity defines only compound or composite
+primary keys, the directive cannot be defined on a leaf type argument and must be defined on a composite type argument.
+- The type of the argument must be equally or more restrictive than the type of the 
+  valid key field, _i.e._, an appropriate subtype, _e.g._, `String!` for `String` or `[Int!]!` for `[Int]`.
+- Naming is irrelevant. If the value of the argument is valid, it will be propagated to the key field.
+
+A composition error will occur if any of these rules are broken.
+
+##### Examples:
+1. Valid: Argument name `id` and key field name `id` match.
+```graphql
+type Mutation {
+    updateEmail(
+      id: ID! @openfed__prerequisite(resolveEntity: { typeName: "User", fields: "id" })
```

Comment:
If I'm understanding correctly, you have to provide the key fields, and then the rest of the fields get fetched and provided for you by the Router? Do the non-ID fields get stripped out of the composed schema? Otherwise I imagine this would be confusing and people would try passing in these fields.

I'm not totally following the use case — if I want to update the email, why would I fetch the email from the User subgraph? Don't I want to overwrite that value with a new email?

Reply:
You want to update an email on a user. How do you know that user you're trying to update with the input is a valid user?